experiments/browser_based_querying/src/urlCompression.ts

```diff
@@ -0,0 +1,361 @@
+// Decoding tables:
+// "-" => " "
+// "--_0" thru "--_9", then "--_a" thru "--_z", then "--_A" thru "--_Z" => 2+ spaces, run-length encoded
+// "--0" thru "--9", then "--a" thru "--z", then "--A" thru "--Z" => "{\n" followed by 2+ spaces, run-length encoded
+// "--*0" thru "--*9", then "--*a" thru "--*z", then "--*A" thru "--*Z" => "}\n" followed by 2+ spaces, run-length encoded
+// N.B.: "---" is a not a legal codeword because it means "space followed by '--' escape sequence"
+//
+// "*l" => "\n"
+// "*L" => ","
+// "**" => '"'
+// "*-" => "-"
+// "*c" => "["
+// "*j" => "]"
+// "*C" => "{"
+// "*J" => "}"
+// "*B" => ":"
+// "*1" => "!"
+// "*2" => "@"
+// "*3" => "#"
+// "*4" => "$"
+// "*5" => "%"
+// "*8" => "*"
+// "*9" => "("
+// "*0" => ")"
+// "*g" => "<"
+// "*G" => ">"
+// "*e" => "="
+// "*o" => "@output"
+// "*O" => "@optional"
+// "*t" => "@tag"
+// "*f" => "@filter"
+// "*T" => "@transform"
+// "*F" => "@fold"
+// "*r" => "@recurse"
+// "*E" => "... on "
+// "*n" => "name:"
+// "*d" => "depth:"
+// "*v" => "value:"
+// "*p" => "op:"
+
+function invertAndStripPrefix(data: Record<string, string>, prefix: string): Record<string, string> {
+  const result: Record<string, string> = {};
+  for (const key in data) {
+    let value = data[key];
+    if (value.startsWith(prefix)) {
+      value = value.substring(prefix.length);
+    }
+    result[value] = key;
+  }
+  return result;
+}
+
+const DIRECTIVE_REPLACEMENTS: Record<string, string> = {
+  '@output': '*o',
+  '@optional': '*O',
+  '@tag': '*t',
+  '@transform': '*T',
+  '@filter': '*f',
+  '@fold': '*F',
+  '@recurse': '*r',
+}
+
+const SPECIAL_CHAR_REPLACEMENTS: Record<string, string> = {
+  '"': '**',
+  '-': '*-',
+  '[': '*c',
+  ']': '*j',
+  '{': '*C',
```

Comment:
We can further compress the common space-prefixed `{` if we have either:
- An encoding token that indicates "remove the following space". This would also improve the compression of poorly-formatted `@` tokens in my other suggestion
- Two separate `{` and ` {` encoding tokens